src/hardware/opl.cpp

```diff
@@ -488,46 +492,92 @@ void OPL::Init(const uint16_t sample_rate)
 	}
 }
 
-void OPL::WriteReg(const uint32_t reg, const uint8_t val)
+void OPL::WriteReg(const io_port_t selected_reg, const uint8_t val)
 {
-	OPL3_WriteRegBuffered(&oplchip, static_cast<uint16_t>(reg), val);
-	if (reg == 0x105)
-		newm = reg & 0x01;
+	OPL3_WriteRegBuffered(&oplchip, selected_reg, val);
+	if (selected_reg == 0x105)
+		newm = selected_reg & 0x01;
 }
 
-uint32_t OPL::WriteAddr(const io_port_t port, const uint8_t val)
+io_port_t OPL::WriteAddr(const io_port_t port, const uint8_t val)
 {
-	uint16_t addr;
-	addr = val;
+	io_port_t addr = val;
 	if ((port & 2) && (addr == 0x05 || newm)) {
 		addr |= 0x100;
 	}
 	return addr;
 }
 
-void OPL::Generate(const mixer_channel_t &chan, const uint16_t frames)
+AudioFrame OPL::RenderFrame()
 {
-	constexpr auto render_frames = 128;
+	static int16_t buf[2] = {};
+	OPL3_GenerateStream(&oplchip, buf, 1);
 
-	int16_t buf[render_frames * 2];
-	float float_buf[render_frames * 2];
+	AudioFrame frame = {};
+	if (adlib_gold) {
+		adlib_gold->Process(buf, 1, &frame[0]);
+	} else {
+		frame.left  = buf[0];
+		frame.right = buf[1];
+	}
+	return frame;
+}
 
-	int remaining = frames;
-	while (remaining > 0) {
-		const auto todo = std::min(remaining, render_frames);
-		OPL3_GenerateStream(&oplchip, buf, todo);
+void OPL::RenderUpToNow()
+{
+	const auto now = PIC_FullIndex();
 
-		if (adlib_gold) {
-			adlib_gold->Process(buf, todo, float_buf);
-			chan->AddSamples_sfloat(todo, float_buf);
-		} else {
-			chan->AddSamples_s16(todo, buf);
-		}
-		remaining -= todo;
+	// Wake up the channel and update the last rendered time datum.
+	if (!channel->is_enabled) {
+		channel->Enable(true);
+		last_rendered_ms = now;
+		return;
 	}
+	// Keep rendering until we're current
+	while (last_rendered_ms < now) {
+		last_rendered_ms += ms_per_frame;
+		fifo.emplace(RenderFrame());
+	}
+}
+
+bool OPL::ChannelCanSleep()
+{
+	for (auto i = 0xb0; i < 0xb9; ++i)
+		if (cache[i] & 0x20 || cache[i + 0x100] & 0x20)
+			return false;
+	return true;
 }
 
-void OPL::CacheWrite(const uint32_t port, const uint8_t val)
+void OPL::AudioCallback(const uint16_t requested_frames)
+{
+	assert(channel);
+
+	//if (fifo.size())
+	//	LOG_MSG("OPL: Queued %2lu cycle-accurate frames", fifo.size());
+
+	auto frames_remaining = requested_frames;
+
+	// First, send any frames we've queued since the last callback
+	while (frames_remaining && fifo.size()) {
+		channel->AddSamples_sfloat(1, &fifo.front()[0]);
+		fifo.pop();
+		--frames_remaining;
+	}
+	// If the queue's run dry, render the remainder and sync-up our time datum
+	while (frames_remaining) {
+		const auto frame = RenderFrame();
+		channel->AddSamples_sfloat(1, &frame[0]);
+		--frames_remaining;
+	}
+	last_rendered_ms = PIC_FullIndex();
+
+	// Maybe idle the channel if the device has been unused for some time
+	constexpr uint16_t three_seconds_of_callbacks = 3 * 1000;
```

Comment:
I noticed that different sound device implementations use different timeout durations for muting the channel. I don't think this makes a whole lot of sense; probably it would be better to introduce a common timeout duration constant that all implementations would then consistently use.

Anyway, maybe not something that you want to address as part of this PR, just something worth mentioning (I think).

Reply:
Yes, it warrants another pass - and a bit more documentation.

In some cases, the last IO port write is all we need to know if the channel should be kept alive or not. The challenge is finding the interfaces to the card that allow tones to be persisted for some number of seconds.

If I recall, one of the PS/1 Audio's registers maximum values can persist a tone for minutes, without any further writes needed to the IO port. That's where the extra silence-tally comes in.  If the card is playing silence (even if that register was populated), well, we can still shutdown the channel in the meantime until the game makes another write.

I think the safest and best is to just do both checks (last IO port + sequence silence), in which case, we can safely sleep the channel very quickly.

Reply:
I think I'd prefer a simple and generic solution: just introduce an accumulator that simply accumulates the output no matter what, and you reset it at say every second. So it's a double that you add the sample values to, and you set it to zero at 1000ms intervals. Then you just simply need to count if the accumulator stayed at absolute zero for N times in a row, where N is time sleep timeout in seconds.

I think it's quite simple, low-cost, low-complexity, and completely universal — no need to know anything at all about the implementation, it only works by inspecting the audio output (like a real-world effect processor, such as a gate would).

Reply:
Very nice idea! Will try that out next.